File: session/opencv/p4-color/color_detector.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
using namespace cv;
using namespace std;

int main() {
    // Cargar imagen
    Mat img = imread("shapes_and_colors.png");
    if (img.empty()) {
        cerr << "Error: no se pudo cargar la imagen." << endl;
        return -1;
    }

    // Convertir de BGR a HSV
    Mat img_hsv;
    cvtColor(img, img_hsv, COLOR_BGR2HSV);

    // Definir rango de tono (hue)
    Scalar lower(7, 100, 100);
    Scalar upper(20, 255, 255);

    cout << "Lower: " << lower << endl;
    cout << "Upper: " << upper << endl;

    // Crear máscara en base al rango de color
    Mat mask;
    inRange(img_hsv, lower, upper, mask);
    /*
        void cv::inRange(
            InputArray src,   // [argin] Imagen de entrada (Mat) en cualquier tipo soportado
            InputArray lowerb,// [argin] Límite inferior del rango (Scalar o Mat)
            InputArray upperb,// [argin] Límite superior del rango (Scalar o Mat)
            OutputArray dst   // [argout] Imagen binaria de salida (Mat) con valores 0 o 255
        );
        ----------------
        - Compara cada píxel de la imagen 'img_hsv' con los límites 'lower' y 'upper'.
        - Si el píxel está dentro del rango HSV especificado, el valor en 'mask' será 255.
        - Si está fuera del rango, el valor será 0.
        Resultado: imagen binaria (blanco = dentro del rango, negro = fuera).
    */

    // Aplicar la máscara sobre la imagen original
    Mat result;
    bitwise_and(img, img, result, mask);
    /*
        void cv::bitwise_and(
            InputArray src1,      // [argin] Primera imagen de entrada
            InputArray src2,      // [argin] Segunda imagen de entrada
            OutputArray dst,      // [argout] Imagen de salida donde se aplicó el AND bit a bit
            InputArray mask = noArray() // [argin] Máscara opcional; sólo se procesan los píxeles distintos de 0
        );
        --------------------
        - Realiza una operación AND bit a bit entre las dos imágenes de entrada (aquí, 'img' y 'img').
        - Sólo los píxeles donde la máscara 'mask' sea distinta de cero (255) se conservan.
        - Los píxeles donde la máscara vale 0 se ponen a negro.
        Resultado: imagen filtrada que muestra sólo los colores dentro del rango definido.
    */

    // Mostrar resultados
    imshow("original", img);
    imshow("mask", mask);
    imshow("filtered", result);
    waitKey(0);
    destroyAllWindows();

    return 0;
}

File: session/opencv/p4-color/color_detector.py
# cap = cv2.VideoCapture(2)
# cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
# cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)
# while True:
    # _, frame = cap.read()
	# hsv_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
	# height, width, _ = frame.shape
    # cx = int(width / 2)
    # cy = int(height / 2)
    # Pick pixel value
    # pixel_center = hsv_frame[cy, cx]
    # hue_value = pixel_center[0]


# import cv2
# import numpy as np

# def nop(x):
    # pass
# cv2.namedWindow("frame")
# cv2.createTrackbar("h", "frame", 0,179, nop)
# cv2.createTrackbar("s", "frame", 255,255, nop)
# cv2.createTrackbar("v", "frame", 255,255, nop)

# while True:
    # h = cv2.getTrackbarPos("h", "frame")
    # s = cv2.getTrackbarPos("s", "frame")
    # v = cv2.getTrackbarPos("v", "frame")

    # img_hsv = np.full((480, 640, 3), (h, s, v), dtype=np.uint8)
    # img = cv2.cvtColor(img_hsv, cv2.COLOR_HSV2BGR)

    # cv2.imshow("frame", img)
    # if cv2.waitKey(1)==27:
        # break

# cv2.destroyAllWindows()



import cv2
import numpy as np

# Load image
img = cv2.imread("shapes_and_colors.png")
img_hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)

#  Choose target hue (for example: red = 0, green = 60, blue = 120)
target_hue = 5  # green

# Define hue tolerance (range)
lower = np.array([7, 100, 100])
upper = np.array([20, 255, 255])
print(lower)
print(upper)

#  Create mask based on hue range
mask = cv2.inRange(img_hsv, lower, upper)

#  Apply mask 
result = cv2.bitwise_and(img, img, mask=mask)

# Show results
cv2.imshow("original", img)
cv2.imshow("mask", mask)
cv2.imshow("filtered", result)
cv2.waitKey(0)
cv2.destroyAllWindows()
